services/pull/merge.go

```diff
@@ -30,8 +31,82 @@ import (
 	"code.gitea.io/gitea/modules/timeutil"
 	asymkey_service "code.gitea.io/gitea/services/asymkey"
 	issue_service "code.gitea.io/gitea/services/issue"
+
+	"github.com/unknwon/com"
 )
 
+// GetDefaultMergeMessage returns default message used when merging pull request
+func GetDefaultMergeMessage(baseGitRepo *git.Repository, pr *models.PullRequest, mergeStyle repo_model.MergeStyle) string {
+	if err := pr.LoadHeadRepo(); err != nil {
+		log.Error("LoadHeadRepo[%d]: %v", pr.HeadRepoID, err)
+		return ""
+	}
+	if err := pr.LoadIssue(); err != nil {
+		log.Error("Cannot load issue %d for PR id %d: Error: %v", pr.IssueID, pr.ID, err)
+		return ""
+	}
+	if err := pr.LoadBaseRepo(); err != nil {
+		log.Error("LoadBaseRepo: %v", err)
+		return ""
+	}
+
+	isExternalTracker := pr.BaseRepo.UnitEnabled(unit.TypeExternalTracker)
+	issueReference := "#"
+	if isExternalTracker {
+		issueReference = "!"
+	}
+
+	if mergeStyle != "" {
+		templateFilepath := fmt.Sprintf(".gitea/default_merge_message/%s_TEMPLATE.md", strings.ToUpper(string(mergeStyle)))
+		commit, err := baseGitRepo.GetBranchCommit(pr.BaseRepo.DefaultBranch)
+		if err != nil {
+			log.Error("GetBranchCommit: %v", err)
+			return ""
+		}
+		templateContent, ok := commit.GetFileContent(templateFilepath)
+		if ok {
+			vars := map[string]string{
+				"BaseRepoOwner":          pr.BaseRepo.OwnerName,
+				"BaseRepoName":           pr.BaseRepo.Name,
+				"BaseBranch":             pr.BaseBranch,
+				"HeadRepoOwner":          pr.HeadRepo.OwnerName,
+				"HeadRepoName":           pr.HeadRepo.Name,
+				"HeadBranch":             pr.HeadBranch,
+				"PullRequestTitle":       pr.Issue.Title,
+				"PullRequestDescription": pr.Issue.Content,
+				"PullRequestPoster":      pr.Issue.Poster.Name,
+				"PullRequestIndex":       strconv.FormatInt(pr.Index, 10),
+				"IssueReferenceChar":     issueReference,
+			}
+			refs, err := pr.ResolveCrossReferences()
+			if err == nil {
+				closeIssueIndexes := make([]string, 0, len(refs))
+				for _, ref := range refs {
+					if ref.RefAction == references.XRefActionCloses {
+						closeIssueIndexes = append(closeIssueIndexes, fmt.Sprintf("close %s%d", issueReference, ref.Issue.Index))
+					}
+				}
+				if len(closeIssueIndexes) > 0 {
+					vars["ClosedIssueIndexes"] = strings.Join(closeIssueIndexes, ", ")
+				}
```

Comment:
I don't think it is a good idea to only provide this when issues are actually closed.
This will lead to unexpected behavior when using the default message for PRs that don't close an issue.
Instead, the message should be initialized with the empty string in that case.